rooted GED

cgoliver opened the pull request:
Added an option to graph_edit_distance called `roots` which is a tuple of nodes, one from each graph.

If specified, the GED computation starts with those nodes matched. This can be used to compare rooted graphs where GED should be compared with respect to the roots. Also leads to faster computation since the roots no longer need to be matched.

pep8speaks commented:
Hello @cgoliver! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`networkx/algorithms/similarity.py`](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py):

> [Line 1017:8](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1017): [E714](https://duckduckgo.com/?q=pep8%20E714) test for object identity should be 'is not'
> [Line 1020:9](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1020): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 1037:12](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1037): [E714](https://duckduckgo.com/?q=pep8%20E714) test for object identity should be 'is not'
> [Line 1047:12](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1047): [E714](https://duckduckgo.com/?q=pep8%20E714) test for object identity should be 'is not'
> [Line 1145:72](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1145): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1145:85](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1145): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1413:1](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1413): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1416:45](https://github.com/networkx/networkx/blob/c88fa2b9d286aab46e2ae67f8eb0690b5e723fa3/networkx/algorithms/similarity.py#L1416): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1005,6 +1013,15 @@ def get_edit_paths(
+        root_u, root_v = roots
+
+        # remove roots from pending
+        del pending_u[pending_u.index(root_u)]
```
would ```pending_u.remove(root_u)``` be more clear?

cgoliver replied:
yes!

dschult commented:
Could you add some tests to make sure the roots functionality continues to work when other changes come in?
Thanks for this!

cgoliver commented:
> Could you add some tests to make sure the roots functionality continues to work when other changes come in?
> Thanks for this!

I added some tests in the docstring, let me know if that is enough.
Glad I could help!

dschult commented:
You also added a timeout feature.  We need to test that too. Nothing in depth, but add test of a short timeout.

And roots should be a 2-tuple (documented as a tuple).  
Does this give nice clear error if roots is not a 2-tuple? What if not nodes of G? etc.

cgoliver commented:
> You also added a timeout feature. We need to test that too. Nothing in depth, but add test of a short timeout.
> 
> And roots should be a 2-tuple (documented as a tuple).
> Does this give nice clear error if roots is not a 2-tuple? What if not nodes of G? etc.

Ah yes sorry about that, first time contributing.

I added asserts for the tuples being the correct size, and the nodes belonging in their respective graphs. Can't think of anything else for the moment.

I put a test for the timeout in the docstring, I thought the timeout was too small for its own PR so I thought I would put it in here, but feel free to remove it.

Sorry for the issues!

dschult commented:
No problem with the issues... :}

Could you make the timeout in your examples shorter -- so the tests don't take so long, and so the result doesn't depend on the machine doing the tests. Maybe 0.5 instead of 5?

dschult commented:
I was having trouble with the tests on my slow machine and since CI testing is done on a variety of machines I moved testing timeout to the ```tests/test_similarity.py``` file.

I also caught a bug when timeout is 0, that would ignore timeout because of ```if timeout:```
To check if timeout is input we should use:  ```if timeout is not None:``` instead of ```if timeout:```.

I also changed the "assert" to raise a NetworkXError instead of the more generic AssertionError.

I think this is ready to go.

rossbar commented on networkx/algorithms/similarity.py:
```diff
@@ -1005,6 +1029,17 @@ def get_edit_paths(
 
+    initial_cost = 0
+    if roots:
+        assert len(roots) == 2, "Roots must be a 2-tuple."
+        root_u, root_v = roots
+        assert root_u in pending_u, "Root node not in graph."
+        assert root_v in pending_v, "Root node not in graph."
```
```suggestion
        if len(roots) != 2:
            ValueError("roots must be a 2-tuple.")
        root_u, root_v = roots
        if (root_u not in pending_u) or (root_v not in pending_v):
            raise ValueError("root node not in graph.")
```
I think it's generally best-practice to raise specific exceptions rather than have assertions in the code - just a suggestion though! Perhaps there's a better nx-specific exception than ValueError for the second case?

dschult replied:
Thanks for this!  I was wondering about assert vs package specific Exception. 
There is a ```NetworkXNodeNotFound``` for the second case.

But should the first be a NetworkXError or a ValueError? What's best practice? We've put NetworkXError for most errors we flag.  If it is a ValueError couldn't we omit it and let the Python error raise on the next line?

cgoliver replied:
seems reasonable to me, probably `nx.NetworkXError` would be better, judging from @dschult 's previous commit.

rossbar replied:
> But should the first be a NetworkXError or a ValueError? What's best practice? We've put NetworkXError for most errors we flag. If it is a ValueError couldn't we omit it and let the Python error raise on the next line?

Good point - Python will emit a ValueError on its own if the unpacking fails. The question becomes whether you want a specific message for the exception, i.e. `roots must be a 2-tuple` vs. `not enough values to unpack`. The most "Pythonic" approach might be:

```python
try:
    root_u, root_v = roots
except ValueError:
    raise ValueError('roots must be a 2-tuple')
```

but IMO any of the proposed solutions would be fine.

dschult replied:
I'm going to let the Python error stand for 'not enough values to unpack' since the code ```root_u, root_v = roots``` describes pretty well which input variable is causing the trouble.

dschult commented:
I approve this PR...

rossbar commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -43,6 +43,22 @@ def setup_class(cls):
+        assert graph_edit_distance(G0, G1, roots=(0, 1), timeout=5) == 8
+        assert graph_edit_distance(G0, G1, roots=(0, 1), timeout=0.0001) is None
+        # test raise on 0 timeout
+        pytest.raises(nx.NetworkXError, graph_edit_distance, G0, G1, timeout=0)
```
Just a question: does anyone know of a good way to test things like timeouts? I don't know of any pattern for reliably *testing* things that depend on dynamic metrics like execution time.

cgoliver replied:
not sure if this is what you're asking.. but on my end I was checking that the actual execution time does not exceed the timeout value.

rossbar replied:
Right, but in order for the test to validate that the timeout keyword is having the intended effect for long-running `foo`, it is necessary that `foo(timeout=None)` will always take longer than `timeout`. This can get tricky as execution times depend on all sorts of things outside of the context of the test itself (platform, hardware, etc.) I'm not sure how to design a good test that is insensitive to all of these factors and reliably tests the timeout itself. Anyways - not a blocker here, but just curious if any has any good ideas!

dschult replied:
I thought that by setting a very very short timeout the first time the code runs the "prune" function that time will already have passed and no progress would have been made so there isn't any solution to return and ```None``` gets returned.  Probably not the best way, but it at least tests that the output is different from the output with a longer timeout.  I picked timeout=5 for the other cases because that should be long enough to get the full answer -- so that just tests that the timeout input is a valid input.

Did any of the travis-ci runs not pass these tests yet?  :}

rossbar replied:
> Did any of the travis-ci runs not pass these tests yet? :}

Nope, tests are all green (for me locally and on CI) - this was more a high-level question on my part, just wondering if anyone had any insight into best-practices for testing things that depend on execution times. The current tests seem totally fine to me, sorry for dragging us off into the weeds :)

rossbar commented:
Once again, thanks @cgoliver for the nice enhancement and @dschult for the excellent review and intersesting discussions!
